File: src/tools/detect-controller-type.ts
import { execSync } from 'child_process';
import { existsSync } from 'fs';
import { join } from 'path';

export interface ControllerTypeInfo {
  resource: string;
  type: 'direct' | 'terraform' | 'unknown';
  location: string | null;
  migration_needed: boolean;
  has_direct_types: boolean;
  has_terraform_types: boolean;
  has_proto: boolean;
  proto_location: string | null;
  service: string | null;
  version: string | null;
}

export function detectControllerType(repoPath: string, resource: string): ControllerTypeInfo {
  const resourceLower = resource.toLowerCase();

  // Check for direct controller types (apis/{service}/{version}/*_types.go)
  const directTypesCmd = `find apis/ -name "*${resourceLower}*_types.go" 2>/dev/null || true`;
  const directTypesFiles = execSync(directTypesCmd, { cwd: repoPath, encoding: 'utf-8' })
    .trim()
    .split('\n')
    .filter(Boolean);

  // Check for Terraform-based types (pkg/clients/generated/)
  const tfTypesCmd = `find pkg/clients/generated/ -name "*${resourceLower}*types.go" 2>/dev/null || true`;
  const tfTypesFiles = execSync(tfTypesCmd, { cwd: repoPath, encoding: 'utf-8' })
    .trim()
    .split('\n')
    .filter(Boolean);

  const hasDirectTypes = directTypesFiles.length > 0;
  const hasTerraformTypes = tfTypesFiles.length > 0;

  let type: 'direct' | 'terraform' | 'unknown' = 'unknown';
  let location: string | null = null;
  let service: string | null = null;
  let version: string | null = null;

  if (hasDirectTypes) {
    type = 'direct';
    location = directTypesFiles[0];

    // Parse: apis/{service}/{version}/{resource}_types.go
    const parts = location.split('/');
    if (parts.length >= 4) {
      service = parts[1];
      version = parts[2];
    }
  } else if (hasTerraformTypes) {
    type = 'terraform';
    location = tfTypesFiles[0];

    // Parse: pkg/clients/generated/apis/{service}/{version}/{resource}_types.go
    const parts = location.split('/');
    if (parts.length >= 6) {
      service = parts[4];
      version = parts[5];
    }
  }

  // Check for proto definitions
  let hasProto = false;
  let protoLocation: string | null = null;

  if (service) {
    // Check in mockgcp/third_party/googleapis
    const protoCmd = `find mockgcp/third_party/googleapis -path "*/${service}/*" -name "*.proto" 2>/dev/null | head -5 || true`;
    const protoFiles = execSync(protoCmd, { cwd: repoPath, encoding: 'utf-8' })
      .trim()
      .split('\n')
      .filter(Boolean);

    if (protoFiles.length > 0) {
      hasProto = true;
      protoLocation = protoFiles[0];
    }
  }

  return {
    resource,
    type,
    location,
    migration_needed: type === 'terraform',
    has_direct_types: hasDirectTypes,
    has_terraform_types: hasTerraformTypes,
    has_proto: hasProto,
    proto_location: protoLocation,
    service,
    version,
  };
}

File: go/internal/tools/detect_controller_type.go
package tools

import (
	"fmt"
	"os/exec"
	"strings"
)

// ControllerTypeInfo contains information about a resource's controller type
type ControllerTypeInfo struct {
	Resource           string  `json:"resource"`
	Type               string  `json:"type"` // "direct", "terraform", or "unknown"
	Location           *string `json:"location"`
	MigrationNeeded    bool    `json:"migration_needed"`
	HasDirectTypes     bool    `json:"has_direct_types"`
	HasTerraformTypes  bool    `json:"has_terraform_types"`
	HasProto           bool    `json:"has_proto"`
	ProtoLocation      *string `json:"proto_location"`
	Service            *string `json:"service"`
	Version            *string `json:"version"`
}

// DetectControllerType detects if a resource uses direct controller or Terraform
func DetectControllerType(repoPath, resource string) (*ControllerTypeInfo, error) {
	resourceLower := strings.ToLower(resource)

	// Check for direct controller types (apis/{service}/{version}/*_types.go)
	directTypesCmd := fmt.Sprintf("find apis/ -name \"*%s*_types.go\" 2>/dev/null || true", resourceLower)
	cmd := exec.Command("sh", "-c", directTypesCmd)
	cmd.Dir = repoPath
	directOutput, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("failed to search for direct types: %w", err)
	}

	directTypesFiles := strings.Split(strings.TrimSpace(string(directOutput)), "\n")
	directTypesFiles = removeEmpty(directTypesFiles)

	// Check for Terraform-based types (pkg/clients/generated/)
	tfTypesCmd := fmt.Sprintf("find pkg/clients/generated/ -name \"*%s*types.go\" 2>/dev/null || true", resourceLower)
	cmd = exec.Command("sh", "-c", tfTypesCmd)
	cmd.Dir = repoPath
	tfOutput, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("failed to search for terraform types: %w", err)
	}

	tfTypesFiles := strings.Split(strings.TrimSpace(string(tfOutput)), "\n")
	tfTypesFiles = removeEmpty(tfTypesFiles)

	hasDirectTypes := len(directTypesFiles) > 0
	hasTerraformTypes := len(tfTypesFiles) > 0

	controllerType := "unknown"
	var location *string
	var service *string
	var version *string

	if hasDirectTypes {
		controllerType = "direct"
		location = &directTypesFiles[0]

		// Parse: apis/{service}/{version}/{resource}_types.go
		parts := strings.Split(*location, "/")
		if len(parts) >= 4 {
			service = &parts[1]
			version = &parts[2]
		}
	} else if hasTerraformTypes {
		controllerType = "terraform"
		location = &tfTypesFiles[0]

		// Parse: pkg/clients/generated/apis/{service}/{version}/{resource}_types.go
		parts := strings.Split(*location, "/")
		if len(parts) >= 6 {
			service = &parts[4]
			version = &parts[5]
		}
	}

	// Check for proto definitions
	hasProto := false
	var protoLocation *string

	if service != nil {
		protoCmd := fmt.Sprintf("find mockgcp/third_party/googleapis -path \"*/%s/*\" -name \"*.proto\" 2>/dev/null | head -5 || true", *service)
		cmd = exec.Command("sh", "-c", protoCmd)
		cmd.Dir = repoPath
		protoOutput, err := cmd.Output()
		if err == nil {
			protoFiles := strings.Split(strings.TrimSpace(string(protoOutput)), "\n")
			protoFiles = removeEmpty(protoFiles)
			if len(protoFiles) > 0 {
				hasProto = true
				protoLocation = &protoFiles[0]
			}
		}
	}

	return &ControllerTypeInfo{
		Resource:          resource,
		Type:              controllerType,
		Location:          location,
		MigrationNeeded:   controllerType == "terraform",
		HasDirectTypes:    hasDirectTypes,
		HasTerraformTypes: hasTerraformTypes,
		HasProto:          hasProto,
		ProtoLocation:     protoLocation,
		Service:           service,
		Version:           version,
	}, nil
}
